docs/developer_guidelines.md

```diff
@@ -0,0 +1,147 @@
+# Developer guidelines
+
+## Table of Contents 📜
+
+<!-- vim-markdown-toc GFM -->
+
+* [Commits 🪢](#commits-)
+    * [Message style](#message-style)
+        * [Short](#short)
+        * [Full](#full)
+    * [Commit and message content](#commit-and-message-content)
+* [Documentation in source code 📑](#documentation-in-source-code-)
+* [Coding standards 📚](#coding-standards-)
+    * [C/C++ 🇨](#cc-)
+    * [Python 🐍](#python-)
+    * [Rust 🦀](#rust-)
+    * [JavaScript/TypeScript 🇯🇸 / 🇹🇸](#javascripttypescript---)
+
+<!-- vim-markdown-toc -->
+
+## Commits 🪢
+
+This section outlines general guidelines to follow when making commits and
+writing commit messages.
+
+A lot has been said on this topic, so it is best to refer to external resources
+and then highlight important points.
+
+Resources:
+* [Source control commit guidelines by Embedded Artistry] - The main resource
+  related to the commit content and message content. **Read this first, before
+  reading the rest**.
+* [One Idea is One Commit] - Interesting committing strategy, comparable to
+  atomic commits.
+
+Use the `gitlint` tool to check your commit messages. See
+[tools/gitlint/README.md] for more information.
+
+### Message style
+
+There are two possible styles for the commit messages: short and full.
+
+#### Short
+
+Short commit messages should be around 50 characters long, the maximum is 72.
+
+Use this when the change that you made is either:
+* self-explanatory, for example: `Fix typo in README.md` or
+* there is a GitHub issue that provides sufficient context for the change made.
+
+Mention relevant GitHub issues at the end of the message, for example:
+* `Fix out of bounds bug, closes #123`
```

Comment:
Does this actually work to close the issue? I thought GitHub required it to be on its own line, but maybe that is not true anymore!

Reply:
Yep, it does! It does not matter where in the commit message the issue is mentioned. But to close the issue with the `Closes` keyword the commit should be pushed to the branch set as a default.

If master is set as default and you are regularly committing to the dev, the issues will only be closed after the merge from dev into master is done.